package/MDAnalysis/analysis/density.py

```diff
@@ -180,11 +192,287 @@
+        self._xdim = xdim
+        self._ydim = ydim
+        self._zdim = zdim
+
+    def _prepare(self):
+        coord = self._atomgroup.positions
+        if self._gridcenter is not None:
+            # Issue 2372: padding is ignored, defaults to 2.0 therefore warn
+            if self._padding > 0:
+                msg = ("Box padding (currently set at {0}) "
+                       "is not used in user defined grids.".format(self._padding))
+                warnings.warn(msg)
+                logger.warning(msg)
+            # Generate a copy of smin/smax from coords to later check if the
+            # defined box might be too small for the selection
+            smin = np.min(coord, axis=0)
+            smax = np.max(coord, axis=0)
+            # Overwrite smin/smax with user defined values
+            smin, smax = _set_user_grid(self._gridcenter, self._xdim,
+                                        self._ydim, self._zdim, smin, smax)
+        else:
+            # Make the box bigger to avoid as much as possible 'outlier'. This
+            # is important if the sites are defined at a high density: in this
+            # case the bulk regions don't have to be close to 1 * n0 but can
+            # be less. It's much more difficult to deal with outliers.  The
+            # ideal solution would use images: implement 'looking across the
+            # periodic boundaries' but that gets complicated when the box
+            # rotates due to RMS fitting.
+            smin = np.min(coord, axis=0) - self._padding
+            smax = np.max(coord, axis=0) + self._padding
+        BINS = fixedwidth_bins(self._delta, smin, smax)
+        arange = np.transpose(np.vstack((BINS['min'], BINS['max'])))
+        bins = BINS['Nbins']
+        # create empty grid with the right dimensions (and get the edges)
+        grid, edges = np.histogramdd(np.zeros((1, 3)), bins=bins,
+                                     range=arange, normed=False)
+        grid *= 0.0
+        self._grid = grid
+        self._edges = edges
+        self._arange = arange
+        self._bins = bins
+        self.density = None
+
+    def _single_frame(self):
+        h, _ = np.histogramdd(self._atomgroup.positions,
+                              bins=self._bins, range=self._arange,
+                              normed=False)
+        # reduce (proposed change #2542 to match the parallel version in pmda.density)
+        # return self._reduce(self._grid, h)
+        #
+        # serial code can simply do
+        self._grid += h
+
+    def _conclude(self):
+        # average:
+        self._grid /= float(self.n_frames)
+        density = Density(grid=self._grid, edges=self._edges,
+                          units={'length': "Angstrom"},
+                          parameters={'isDensity': False})
+        density.make_density()
+        self.density = density
+
+    # _reduce is not strictly necessary for the serial version but is necessary for
```

Comment:
So this is more to try to avoid ending up with commented out code that is forgotten about, assuming that #2542 goes ahead, my understanding is that the `_reduce` method code will be added to :class:AnalysisBase. Is keeping this in the meantime necessary or could the above comment in `_single_frame` be sufficient?

Reply:
This particularly `_reduce` method is different from the standard append one 
https://github.com/MDAnalysis/pmda/blob/26eb31e8269124c0d9e1382adcac8007c04681db/pmda/parallel.py#L446-L450 
so I would leave the commented out code.